server/src/test/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinatorTest.java

```diff
@@ -1133,6 +1169,74 @@ public void testUnusedHighRange() throws IOException
     );
   }
 
+  @Test
+  public void testUsedEternitySegmentEternityFilter() throws IOException
+  {
+    coordinator.announceHistoricalSegments(ImmutableSet.of(eternityRangeSegment1));
+
+    Assert.assertEquals(
+        ImmutableSet.of(eternityRangeSegment1),
+        ImmutableSet.copyOf(
+            coordinator.retrieveUsedSegmentsForIntervals(
+                eternityRangeSegment1.getDataSource(),
+                Intervals.ONLY_ETERNITY,
+                Segments.ONLY_VISIBLE
+            )
+        )
+    );
+  }
+
+  @Test
+  public void testUsedEternitySegment2000Filter() throws IOException
+  {
+    coordinator.announceHistoricalSegments(ImmutableSet.of(eternityRangeSegment1));
+
+    Assert.assertEquals(
+        ImmutableSet.of(eternityRangeSegment1),
+        ImmutableSet.copyOf(
+            coordinator.retrieveUsedSegmentsForIntervals(
+                eternityRangeSegment1.getDataSource(),
+                Collections.singletonList(Intervals.of("2000/2030")),
+                Segments.ONLY_VISIBLE
+            )
+        )
+    );
+  }
+
+  @Test
+  public void testUsedHalfEternitySegmentEternityFilter() throws IOException
+  {
+    coordinator.announceHistoricalSegments(ImmutableSet.of(halfEternityRangeSegment1, halfEternityRangeSegment2));
+
+    Assert.assertEquals(
+        ImmutableSet.of(halfEternityRangeSegment1, halfEternityRangeSegment2),
+        ImmutableSet.copyOf(
+            coordinator.retrieveUsedSegmentsForIntervals(
+                halfEternityRangeSegment1.getDataSource(),
+                Intervals.ONLY_ETERNITY,
+                Segments.ONLY_VISIBLE
+            )
+        )
+    );
+  }
+
+  @Test
+  public void testUsedHalfEternitySegment2000Filter() throws IOException
```

Comment:
Took me a while wondering how this pass before realizing that there is a filter done on the result return from sql